tools/rtrclient.c

```diff
@@ -7,35 +7,188 @@
  * Website: http://rtrlib.realmv6.org/
  */
 
-#include <stdlib.h>
-#include <pthread.h>
 #include <arpa/inet.h>
+#include <ctype.h>
+#include <netdb.h>
+#include <pthread.h>
+#include <stdarg.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 #include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
 #include <unistd.h>
+
 #include "rtrlib/rtrlib.h"
 
+enum socket_type {
+	SOCKET_TYPE_TCP,
+#ifdef RTRLIB_HAVE_LIBSSH
+	SOCKET_TYPE_SSH,
+#endif
+};
+
+struct socket_config {
+	struct rtr_socket socket;
+	struct tr_socket tr_socket;
+	enum socket_type type;
+	bool print_pfx_updates;
+	bool print_spki_updates;
+	char *bindaddr;
+	char *host;
+	char *port;
+	char *ssh_username;
+	char *ssh_private_key;
+	char *ssh_host_key;
+};
+
+/* activate update callback */
+bool activate_pfx_update_cb = false;
+bool activate_spki_update_cb = false;
+
+/* print all updates regardless of socket configuration */
+bool print_all_pfx_updates = false;
+bool print_all_spki_updates = false;
+
+struct socket_config **socket_config = NULL;
+size_t socket_count = 0;
+
+pthread_mutex_t stdout_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+/**
+ * @brief Check if argument is NULL and exit if it is
+ */
+static void *check_alloc_ret(void *ptr)
+{
+	if (ptr)
+		return ptr;
+
+	fprintf(stderr, "Memory allocation error\n");
+	exit(-1);
+}
+
+/**
+ * @brief Return value checking malloc wrapper. Exits program on error
+ */
+static inline void *checked_malloc(size_t size)
+{
+	return check_alloc_ret(malloc(size));
+}
+
+/**
+ * @brief Return value checking realloc wrapper. Exits program on error.
+ */
+static inline void *checked_realloc(void *ptr, size_t size)
+{
+	return check_alloc_ret(realloc(ptr, size));
+}
+
+static struct socket_config *extend_socket_config() {
+	socket_config = checked_realloc(
+			socket_config,
+			sizeof(struct socket_config *) * (socket_count + 1));
+	struct socket_config *config = socket_config[socket_count] =
+		checked_malloc(sizeof(struct socket_config));
+
+	memset(config, 0, sizeof(*config));
+	++socket_count;
+
+	return config;
+}
+
+static bool is_numeric(const char *str)
+{
+	for (size_t i = 0; i < strlen(str); ++i) {
+		if (!isdigit(str[i]))
+			return false;
+	}
+	return true;
+}
+
+static bool is_valid_port_number(const char *str)
+{
+	if (!is_numeric(str))
```

Comment:
this check can be omitted by using `strtol` (or even better `strtoul`) instead of `atoi` below, on error it returns `0` which is not a valid port number anyway.

Reply:
unfortunately strtol ignores trailing characters, my check does not.

Reply:
okay